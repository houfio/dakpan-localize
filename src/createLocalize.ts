import { createDakpan } from 'dakpan';
import {
  Component,
  ComponentClass,
  ComponentType,
  createElement,
  ProviderProps,
  ReactElement,
  ReactNodeArray,
  ReactPortal,
  SFCElement,
  StatelessComponent
} from 'react';

import { flattenTranslation } from './flattenTranslation';
import { LocalizeState, TranslateProps, Translations } from './types';

export const createLocalize = (language: string, languages: string[]) => {
  if (!languages.length) {
    throw new Error(`You may not provide 'createLocalize' with an empty 'languages' array`);
  } else if (languages.indexOf(language) === -1) {
    language = languages[0];
  }

  const { Provider, Consumer, withConsumer, context } = createDakpan<LocalizeState>({
    language,
    languages,
    translations: {}
  })({
    setLanguage: (language: string) => () => ({
      language
    })
  });

  return {
    LocalizeProvider: Provider,
    withLocalize: withConsumer,
    withTranslations:
      (withTranslations: Translations) => <P>(component: ComponentType<P>) => (props: P) => createElement(Consumer, {
        children: ({ language, languages, translations }: LocalizeState) => createElement(context.Provider, {
          value: {
            language,
            languages,
            translations: {
              ...translations,
              ...flattenTranslation(withTranslations)
            }
          },
          children: createElement(component, props as P)
        })
      }),
    Translate: ({ id, data = {} }: TranslateProps) => createElement(Consumer, {
      children: ({ language, languages, translations }: LocalizeState) => {
        const index = languages.indexOf(language);
        const translation = translations[id];

        if (!translation) {
          throw new Error(`Translation with id '${id}' not found`);
        } else if (index === -1 || translation.length <= index) {
          throw new Error(`Language '${language}' for translation '${id}' not found`);
        }

        let result = translation[index];

        Object.keys(data).forEach((key) => result = result.replace(new RegExp('\\{' + key + '}', 'g'), data[key]));

        return result;
      }
    })
  };
};
