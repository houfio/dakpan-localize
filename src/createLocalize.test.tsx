import { mount } from 'enzyme';
import * as React from 'react';
import { StatelessComponent } from 'react';

import { createLocalize } from './createLocalize';

it('should override the parent translations', () => {
  const { LocalizeProvider, Translate, withTranslations } = createLocalize('en', ['en']);

  const One: StatelessComponent = withTranslations({
    test: ['hello'],
    hello: ['yes']
  })(() => (
    <>
      <Translate id="test"/>
      <Translate id="hello"/>
      <Two/>
    </>
  ));

  One.displayName = 'One';

  const Two: StatelessComponent = withTranslations({
    test: ['it works!']
  })(() => (
    <>
      <Translate id="test"/>
      <Translate id="hello"/>
    </>
  ));

  Two.displayName = 'Two';

  const wrapper = mount(
    <LocalizeProvider>
      <One/>
    </LocalizeProvider>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should change the language', () => {
  const { LocalizeProvider, Translate, withLocalize, withTranslations } = createLocalize('en', ['en', 'nl']);

  const Component = withTranslations({
    test: {
      message: ['message', 'bericht']
    }
  })(withLocalize((_, { setLanguage }) => ({ setLanguage }))(({ setLanguage }) => (
    <>
      <button onClick={setLanguage.e('nl')}>lang</button>
      <Translate id="test.message"/>
    </>
  )));

  const wrapper = mount(
    <LocalizeProvider>
      <Component/>
    </LocalizeProvider>
  );

  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(wrapper.update()).toMatchSnapshot();
});

it('should replace placeholders with provided data', () => {
  const { LocalizeProvider, Translate, withTranslations } = createLocalize('en', ['en']);

  const Component = withTranslations({
    greeting: ['Hello {name}!']
  })(() => (
    <>
      <Translate id="greeting" data={{ name: 'Steve' }}/>
    </>
  ));

  const wrapper = mount(
    <LocalizeProvider>
      <Component/>
    </LocalizeProvider>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should be possible not to use the hocs', () => {
  const { LocalizeProvider, Translate, TranslationProvider } = createLocalize('en', ['en']);

  const wrapper = mount(
    <LocalizeProvider>
      <TranslationProvider
        translations={{
          test: {
            message: [
              'hello!'
            ]
          }
        }}
      >
        <Translate id="test.message"/>
      </TranslationProvider>
    </LocalizeProvider>
  );

  expect(wrapper).toMatchSnapshot();
});
