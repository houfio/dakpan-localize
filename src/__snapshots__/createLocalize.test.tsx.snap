// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change the language 1`] = `
<Provider>
  <Component>
    <Consumer>
      <WithConsumer>
        <Consumer>
          <Component
            setLanguage={[Function]}
          >
            <button
              onClick={[Function]}
            >
              lang
            </button>
            <Translate
              id="test.message"
            >
              <Consumer>
                message
              </Consumer>
            </Translate>
          </Component>
        </Consumer>
      </WithConsumer>
    </Consumer>
  </Component>
</Provider>
`;

exports[`should change the language 2`] = `
<Provider>
  <Component>
    <Consumer>
      <WithConsumer>
        <Consumer>
          <Component
            setLanguage={[Function]}
          >
            <button
              onClick={[Function]}
            >
              lang
            </button>
            <Translate
              id="test.message"
            >
              <Consumer>
                bericht
              </Consumer>
            </Translate>
          </Component>
        </Consumer>
      </WithConsumer>
    </Consumer>
  </Component>
</Provider>
`;

exports[`should override the parent translations 1`] = `
<Provider>
  <One>
    <Consumer>
      <Component>
        <Translate
          id="test"
        >
          <Consumer>
            hello
          </Consumer>
        </Translate>
        <Translate
          id="hello"
        >
          <Consumer>
            yes
          </Consumer>
        </Translate>
        <Two>
          <Consumer>
            <Component>
              <Translate
                id="test"
              >
                <Consumer>
                  it works!
                </Consumer>
              </Translate>
              <Translate
                id="hello"
              >
                <Consumer>
                  yes
                </Consumer>
              </Translate>
            </Component>
          </Consumer>
        </Two>
      </Component>
    </Consumer>
  </One>
</Provider>
`;

exports[`should replace placeholders with provided data 1`] = `
<Provider>
  <Component>
    <Consumer>
      <Component>
        <Translate
          data={
            Object {
              "name": "Steve",
            }
          }
          id="greeting"
        >
          <Consumer>
            Hello Steve!
          </Consumer>
        </Translate>
      </Component>
    </Consumer>
  </Component>
</Provider>
`;
